server/aws-lsp-codewhisperer/src/language-server/agenticChat/agenticChatController.ts

```diff
@@ -165,141 +153,360 @@ export class AgenticChatController implements ChatHandlers {
-                            triggerType: triggerContext.triggerType,
-                        },
-                    })
+            const currentMessage = currentRequestInput.conversationState?.currentMessage
 
-                    this.#telemetryController.emitStartConversationMetric(params.tabId, metric.metric)
-                }
+            // Process tool uses and update the request input for the next iteration
+            const toolResults = await this.#processToolUses(pendingToolUses)
+            currentRequestInput = this.#updateRequestInputWithToolResults(currentRequestInput, toolResults)
+
+            if (!currentRequestInput.conversationState!.history) {
+                currentRequestInput.conversationState!.history = []
             }
 
-            metric.setDimension('codewhispererCustomizationArn', requestInput.conversationState?.customizationArn)
-            await this.#telemetryController.emitAddMessageMetric(params.tabId, metric.metric)
-
-            this.#telemetryController.updateTriggerInfo(params.tabId, {
-                lastMessageTrigger: {
-                    ...triggerContext,
-                    messageId: response.$metadata.requestId,
-                    followUpActions: new Set(
-                        result.data?.chatResult.followUp?.options
-                            ?.map(option => option.prompt ?? '')
-                            .filter(prompt => prompt.length > 0)
-                    ),
+            currentRequestInput.conversationState!.history.push({
+                userInputMessage: {
+                    content: currentMessage?.userInputMessage?.content,
+                    origin: currentMessage?.userInputMessage?.origin,
+                    userIntent: currentMessage?.userInputMessage?.userIntent,
+                    userInputMessageContext: currentMessage?.userInputMessage?.userInputMessageContext,
+                },
+            })
+
+            currentRequestInput.conversationState!.history.push({
+                assistantResponseMessage: {
+                    content: result.data?.chatResult.body,
+                    toolUses: Object.keys(result.data?.toolUses!).map(k => ({
+                        toolUseId: result.data!.toolUses[k].toolUseId,
+                        name: result.data!.toolUses[k].name,
+                        input: result.data!.toolUses[k].input,
+                    })),
                 },
             })
-            // Save question/answer interaction to chat history
-            if (params.prompt.prompt && session.conversationId && result.data?.chatResult.body) {
-                this.#chatHistoryDb.addMessage(params.tabId, 'cwc', session.conversationId, {
-                    body: params.prompt.prompt,
-                    type: 'prompt' as any,
+        }
+
+        if (iterationCount >= maxIterations) {
+            this.#log('Agent loop reached maximum iterations limit')
```

Comment:
would be nice in the future to have telemetry for this, that way we could see if agentic chat was hitting this too often